### ISSUE TITLE
IValueTaskSource[<T>] very confused on preview7 (IDE only?)

### ISSUE THREAD

**Issue by mgravell** [dotnet/corefx#39717]
IDE: 16.2.0 Preview 4.0

repro: create a new project with a preview 7 global.json, and something like:

``` xml
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <LangVersion>preview</LangVersion>
    <Nullable>enable</Nullable>
```

(a console exe is fine, etc) Now add a .cs file:

``` c#
using System.Threading.Tasks.Sources;

class Foo : IValueTaskSource
{
}
```

Obviously we haven't implemented the interface yet, so let's accept the IDE's ability to implement it:

![image](https://user-images.githubusercontent.com/17328/61786716-3b556600-ae06-11e9-83da-693c528e26bc.png)

The IDE generates:

``` c#
class Foo : IValueTaskSource
{
    public void GetResult(short? token)
    {
        throw new NotImplementedException();
    }

    public ValueTaskSourceStatus? GetStatus(short? token)
    {
        throw new NotImplementedException();
    }

    public void OnCompleted(Action<object?> continuation, object state, short token, ValueTaskSourceOnCompletedFlags flags)
    {
        throw new NotImplementedException();
    }
}
```

which is incorrect; `GetResult` and `GetStatus` take `short`, not `short?`, and the return type is wrong:

``` txt
Severity	Code	Description	Project	File	Line	Suppression State
Error	CS0535	'Foo' does not implement interface member 'IValueTaskSource.GetResult(short)'	ConsoleApp13	C:\Users\marc\source\repos\ConsoleApp13\Foo.cs	4	Active
Error	CS0535	'Foo' does not implement interface member 'IValueTaskSource.GetStatus(short)'	ConsoleApp13	C:\Users\marc\source\repos\ConsoleApp13\Foo.cs	4	Active
```

So, let's fix that manually:

``` c#
class Foo : IValueTaskSource
{
    public void GetResult(short token)
    {
        throw new NotImplementedException();
    }

    public ValueTaskSourceStatus GetStatus(short token)
    {
        throw new NotImplementedException();
    }

    public void OnCompleted(Action<object?> continuation, object? state, short token, ValueTaskSourceOnCompletedFlags flags)
    {
        throw new NotImplementedException();
    }
}
```

But it is still very unhappy:

``` txt
Severity	Code	Description	Project	File	Line	Suppression State
Warning	CS8614	Nullability of reference types in type of parameter 'token' of 'ValueTaskSourceStatus Foo.GetStatus(short token)' doesn't match implicitly implemented member 'ValueTaskSourceStatus IValueTaskSource.GetStatus(short token)'.	ConsoleApp13	C:\Users\marc\source\repos\ConsoleApp13\Foo.cs	11	Active
Warning	CS8614	Nullability of reference types in type of parameter 'flags' of 'void Foo.OnCompleted(Action<object?> continuation, object? state, short token, ValueTaskSourceOnCompletedFlags flags)' doesn't match implicitly implemented member 'void IValueTaskSource.OnCompleted(Action<object?> continuation, object? state, short token, ValueTaskSourceOnCompletedFlags flags)'.	ConsoleApp13	C:\Users\marc\source\repos\ConsoleApp13\Foo.cs	16	Active
Warning	CS8614	Nullability of reference types in type of parameter 'token' of 'void Foo.GetResult(short token)' doesn't match implicitly implemented member 'void IValueTaskSource.GetResult(short token)'.	ConsoleApp13	C:\Users\marc\source\repos\ConsoleApp13\Foo.cs	6	Active
Warning	CS8614	Nullability of reference types in type of parameter 'token' of 'void Foo.OnCompleted(Action<object?> continuation, object? state, short token, ValueTaskSourceOnCompletedFlags flags)' doesn't match implicitly implemented member 'void IValueTaskSource.OnCompleted(Action<object?> continuation, object? state, short token, ValueTaskSourceOnCompletedFlags flags)'.	ConsoleApp13	C:\Users\marc\source\repos\ConsoleApp13\Foo.cs	16	Active
```

and if we look at the definition of `IValueTaskSource` using the "from metadata" view (F12 on the interface), it gets even more confused:

``` c#
using System.Runtime.CompilerServices;

namespace System.Threading.Tasks.Sources
{
    [NullableContextAttribute(2)]
    public interface IValueTaskSource
    {
        void GetResult(short? token);
        ValueTaskSourceStatus? GetStatus(short? token);
        void OnCompleted([NullableAttribute(new[] { 1, 2 })] Action<object?> continuation, object? state, short? token, ValueTaskSourceOnCompletedFlags? flags);
    }
}
```

this *does not match* the [actual definition](https://github.com/dotnet/corefx/blob/b2097cbdcb26f7f317252334ddcce101a20b7f3d/src/Common/src/CoreLib/System/Threading/Tasks/Sources/IValueTaskSource.cs#L45-L61) **nor** does it match what it generated method stubs for earlier **nor** does it match what the compiler is looking for...

`IValueTaskSource<T>` gets confused in very similar ways.

**Comment by mgravell** [comment]
Note that these problems appear to be limited to the IDE; at the command-line (`dotnet build`) the compiler works correctly and does not generate the warnings that the IDE generates.

**Comment by mgravell** [comment]
It is possible that this just needs 16.3 preview 1, which [is delayed (thread)](https://twitter.com/Pegazux/status/1153838867138666497)

**Comment by stephentoub** [comment]
cc: @cston, @jcouv

**Comment by jcouv** [comment]
This looks like the bug we had for a short time where we would decode value types incorrectly. I think PR https://github.com/dotnet/roslyn/pull/36784 was the fix (in 16.3p1). @cston can confirm

**Comment by cston** [comment]
Yes, this looks like the issue fixed in https://github.com/dotnet/roslyn/pull/36784.

**Comment by stephentoub** [comment]
Thanks

**Comment by mgravell** [comment]
confirm fixed in 16.3.0
